src/object/cli_shard.c

```diff
@@ -1682,22 +1708,36 @@ obj_shard_query_recx_post(struct obj_query_key_cb_args *cb_args,
 			goto re_check;
 	}
 
-	end[0] = recx[0].rx_idx + recx[0].rx_nr;
-	end[1] = recx[1].rx_idx + recx[1].rx_nr;
+	end[0] = DAOS_RECX_END(recx[0]);
+	end[1] = DAOS_RECX_END(recx[1]);
+	result_end = DAOS_RECX_PTR_END(result_recx);
 	if (get_max) {
-		if (end[0] > end[1])
-			*result_recx = recx[0];
-		else
-			*result_recx = recx[1];
+		if (end[1] > result_end ||
+		    end[0] > result_end) {
+			if (end[0] > end[1])
+				*result_recx = recx[0];
+			else
+				*result_recx = recx[1];
+		}
 	} else {
-		if (recx[0].rx_nr == 0)
-			*result_recx = recx[1];
-		else if (recx[1].rx_nr == 0)
-			*result_recx = recx[0];
-		else if (end[0] > end[1])
-			*result_recx = recx[1];
-		else
-			*result_recx = recx[0];
+		if (end[0] == 0) {
+			if (result_end == 0 ||
+			    (end[1] != 0 && end[1] < result_end))
+				*result_recx = recx[1];
+		} else if (end[1] == 0) {
+			if (result_end == 0 || end[0] < result_end)
+				*result_recx = recx[0];
+		} else if (end[0] > end[1]) {
+			if (result_end == 0 || end[1] < result_end)
+				*result_recx = recx[1];
+			else if (end[0] < result_end)
+				*result_recx = recx[0];
+		} else {
+			if (result_end == 0 || end[0] < result_end)
+				*result_recx = recx[0];
+			else if (end[1] < result_end)
+				*result_recx = recx[1];
+		}
```

Comment:
yes agree it is not very easy to understand. the get_min case is more complex than get_max, because need to consider the case that if replied recx[0] or/and recx[1] is [0, 0] in that case need to skip that replied recx, that's why check (end[0] == 0) and (end[1] == 0).
"it seems that the "result_recx" is always set as the one that is smaller and next to it"
you mean the logic can be simplified?

Reply:
No, I just try to understand whether the logic is equal to that.